File: mcp/project-agent-node/src/graph.ts
// Project dependency graph loader + lookups.
// Reads PROJECTS_GRAPH.yaml from the workspace root. Used by inbox
// broadcast tools to fan out messages to every dependent project.

import { existsSync, readFileSync, statSync } from "node:fs";
import { join } from "node:path";
import { parse } from "yaml";

interface ProjectEntry {
  description?: string;
  depends_on?: string[];
}

interface GraphFile {
  projects?: Record<string, ProjectEntry>;
}

export class ProjectGraph {
  private graphPath: string;
  private dependsOn = new Map<string, string[]>();   // project → its upstream deps
  private servedBy = new Map<string, string[]>();    // project → its downstream dependents (inverse)
  private projects: string[] = [];
  private loadedMtime: number | null = null;

  constructor(private projectsRoot: string) {
    this.graphPath = join(projectsRoot, "PROJECTS_GRAPH.yaml");
    this.reload();
  }

  /** Reload from disk. Idempotent. Returns true if anything changed. */
  reload(): boolean {
    if (!existsSync(this.graphPath)) {
      const had = this.projects.length > 0;
      this.dependsOn.clear();
      this.servedBy.clear();
      this.projects = [];
      this.loadedMtime = null;
      return had;
    }

    const mtime = statSync(this.graphPath).mtimeMs;
    if (this.loadedMtime !== null && mtime === this.loadedMtime) return false;

    const raw = readFileSync(this.graphPath, "utf8");
    const data = parse(raw) as GraphFile | null;
    const projects = data?.projects ?? {};

    this.dependsOn.clear();
    this.servedBy.clear();
    this.projects = Object.keys(projects);

    for (const [name, entry] of Object.entries(projects)) {
      const deps = entry.depends_on ?? [];
      this.dependsOn.set(name, deps);
      for (const dep of deps) {
        const arr = this.servedBy.get(dep) ?? [];
        arr.push(name);
        this.servedBy.set(dep, arr);
      }
    }

    this.loadedMtime = mtime;
    return true;
  }

  /** Projects this project DEPENDS ON (its upstream). */
  listDependencies(project: string): string[] {
    this.reload();
    return [...(this.dependsOn.get(project) ?? [])];
  }

  /** Projects that DEPEND ON this project (its downstream — fanout targets). */
  listDependents(project: string): string[] {
    this.reload();
    return [...(this.servedBy.get(project) ?? [])];
  }

  /** All projects declared in the graph. */
  listProjects(): string[] {
    this.reload();
    return [...this.projects];
  }

  hasGraph(): boolean {
    this.reload();
    return this.projects.length > 0;
  }
}

File: mcp/project-agent-rs/src/graph.rs
// Project dependency graph loader + lookups.
// Reads PROJECTS_GRAPH.yaml from the workspace root. Used by inbox
// broadcast tools to fan out messages to every dependent project.

use anyhow::Result;
use serde::Deserialize;
use std::collections::HashMap;
use std::path::{Path, PathBuf};
use std::sync::Mutex;

#[derive(Debug, Deserialize, Default)]
struct ProjectEntry {
    #[serde(default)]
    depends_on: Vec<String>,
    #[allow(dead_code)]
    description: Option<String>,
}

#[derive(Debug, Deserialize, Default)]
struct GraphFile {
    #[serde(default)]
    projects: HashMap<String, ProjectEntry>,
}

struct GraphCache {
    depends_on: HashMap<String, Vec<String>>,
    served_by: HashMap<String, Vec<String>>,
    projects: Vec<String>,
    mtime: Option<std::time::SystemTime>,
}

pub struct ProjectGraph {
    graph_path: PathBuf,
    cache: Mutex<GraphCache>,
}

impl ProjectGraph {
    pub fn new(projects_root: &Path) -> Self {
        let graph_path = projects_root.join("PROJECTS_GRAPH.yaml");
        let graph = ProjectGraph {
            graph_path,
            cache: Mutex::new(GraphCache {
                depends_on: HashMap::new(),
                served_by: HashMap::new(),
                projects: Vec::new(),
                mtime: None,
            }),
        };
        let _ = graph.reload();
        graph
    }

    /// Reload from disk. Idempotent. Silent on missing file (clears state).
    fn reload(&self) -> Result<()> {
        let mut cache = self.cache.lock().unwrap();

        if !self.graph_path.exists() {
            cache.depends_on.clear();
            cache.served_by.clear();
            cache.projects.clear();
            cache.mtime = None;
            return Ok(());
        }

        let mtime = std::fs::metadata(&self.graph_path)?.modified().ok();
        if cache.mtime == mtime && cache.mtime.is_some() {
            return Ok(()); // unchanged
        }

        let raw = std::fs::read_to_string(&self.graph_path)?;
        let data: GraphFile = serde_yaml::from_str(&raw).unwrap_or_default();

        cache.depends_on.clear();
        cache.served_by.clear();
        cache.projects = data.projects.keys().cloned().collect();

        for (name, entry) in &data.projects {
            cache.depends_on.insert(name.clone(), entry.depends_on.clone());
            for dep in &entry.depends_on {
                cache
                    .served_by
                    .entry(dep.clone())
                    .or_insert_with(Vec::new)
                    .push(name.clone());
            }
        }

        cache.mtime = mtime;
        Ok(())
    }

    pub fn list_dependencies(&self, project: &str) -> Vec<String> {
        let _ = self.reload();
        let cache = self.cache.lock().unwrap();
        cache.depends_on.get(project).cloned().unwrap_or_default()
    }

    pub fn list_dependents(&self, project: &str) -> Vec<String> {
        let _ = self.reload();
        let cache = self.cache.lock().unwrap();
        cache.served_by.get(project).cloned().unwrap_or_default()
    }

    #[allow(dead_code)]
    pub fn list_projects(&self) -> Vec<String> {
        let _ = self.reload();
        let cache = self.cache.lock().unwrap();
        cache.projects.clone()
    }

    pub fn has_graph(&self) -> bool {
        let _ = self.reload();
        let cache = self.cache.lock().unwrap();
        !cache.projects.is_empty()
    }
}
